### PR TITLE
Unified naming for float16 type.

### PR DESCRIPTION
**Pull request by mzient**
Signed-off-by: Michal Zientkiewicz <michalz@nvidia.com>

#### Why we need this PR?
*Pick one*
- Refactoring to improve handling of 16-bit floating point numbers.

#### What happened in this PR?
* Add `#ifdef` guarded typedefs for float16
* Add `#ifnfdef` guardded inclusion of `half.hpp`
* Remove `normalize_t`
* Change all uses of `float16_cpu` to `float16`.
* Unify cpu/gpu variants of DALI_TYPE_SWITCH_WITH_FP16
* Add `half.hpp` to dali python wheel
* Covered by existing tests.

**JIRA TASK**: [DALI-398]

### CODE REVIEW

**Comment by mzient** [comment]
!build

File: include/dali/core/cuda_utils.h [hunk 1]
```diff
@@ -21,8 +21,10 @@
 #include "dali/core/host_dev.h"
 #include "dali/core/dynlink_cuda.h"
 #include "dali/core/cuda_error.h"
+#ifndef __CUDA_ARCH__
+#include "dali/util/half.hpp"
+#endif
```
Comment (szalpal) [inline include/dali/core/cuda_utils.h]: Do we need the forward declaration, if we include it directly?

Comment (mzient) [reply include/dali/core/cuda_utils.h]: I don't think so.

File: dali/kernels/slice/slice_flip_normalize_permute_gpu.h [hunk 2]
```diff
@@ -102,7 +102,7 @@ class SliceFlipNormalizePermutePadGPU {
       norm_add_cpu[i] = -mean_data[i] * inv_stddev_data[i];
       norm_mul_cpu[i] = inv_stddev_data[i];
     }
-    unsigned normalization_dim;
+    unsigned normalization_dim = Dims + 1;
```
Comment (jantonguirao) [inline dali/kernels/slice/slice_flip_normalize_permute_gpu.h]: was it a bug?

File: include/dali/core/cuda_utils.h [hunk 3]
```diff
@@ -21,20 +21,18 @@
...
 #include "dali/core/dynlink_cuda.h"
 #include "dali/core/cuda_error.h"
-
-// For the CPU we use half_float lib and float16_cpu type
-namespace half_float {
-
-class half;
-
-}
+#ifndef __CUDA_ARCH__
```
Comment (jantonguirao) [inline include/dali/core/cuda_utils.h]: perhaps float16 deserves its own header file separated from cuda utils?

File: include/dali/core/cuda_utils.h [hunk 4]
```diff
@@ -21,20 +21,18 @@
...
 
 // For the GPU
-typedef __half float16;
-// For the CPU
-typedef half_float::half float16_cpu;
+#ifdef __CUDA_ARCH__
+using float16 = __half;
+#else
+using float16 = half_float::half;
+#endif
```
Comment (jantonguirao) [inline include/dali/core/cuda_utils.h]: I am missing the changes in old type switches?

File: dali/kernels/slice/slice_flip_normalize_permute_gpu.h (hunk 2 above)
Comment (mzient) [reply dali/kernels/slice/slice_flip_normalize_permute_gpu.h]: Not really, but the compiler generated a warning and CI failed :(

File: include/dali/core/cuda_utils.h (hunk 4 above)
Comment (mzient) [reply include/dali/core/cuda_utils.h]: I changed them - but now we don't need separate switch for _CPU and _GPU - I'll unify them.

**Comment by mzient** [comment]
!build

File: include/dali/core/cuda_utils.h (hunk 3 above)
Comment (mzient) [reply include/dali/core/cuda_utils.h]: It does, but maybe let's make it another PR and move it to core then.

File: include/dali/core/cuda_utils.h (hunk 4 above)
Comment (jantonguirao) [reply include/dali/core/cuda_utils.h]: Ok, then let's make a follow PR where we move this to core and remove the old switches (_CPU, _GPU)

**Pull request merged**